packages/synapse-react-client/src/components/doi/CreateOrUpdateDoiModal.tsx

```diff
@@ -0,0 +1,324 @@
+import {
+  convertToEntityType,
+  entityTypeToFriendlyName,
+  isTable,
+  isVersionableEntity,
+} from '@/utils/functions/EntityTypeUtils'
+import { HelpTwoTone } from '@mui/icons-material'
+import {
+  Alert,
+  Button,
+  InputLabel,
+  MenuItem,
+  Select,
+  Typography,
+} from '@mui/material'
+import { DoiResourceTypeGeneral, V2Doi } from '@sage-bionetworks/synapse-client'
+import { EntityType } from '@sage-bionetworks/synapse-types'
+import { isEmpty } from 'lodash-es'
+import { useEffect, useState } from 'react'
+
+type CreateOrUpdateDoiModalProps = {
+  /** Whether the dialog is open */
+  open: boolean
+  /** Callback to call when the dialog is closed */
+  onClose: () => void
+  /** The type of object */
+  objectType: V2Doi['objectType']
+  /** The ID of the object */
+  objectId: string
+  /** The optional version number of the object used to populate the form. */
+  defaultVersionNumber?: number
+}
+
+type DoiFormData = {
+  titles?: string[]
+  creators?: string[]
+  resourceTypeGeneral?: string
+  publicationYear?: number
+}
+
+function getSuggestedResourceTypeGeneral(
+  type: EntityType,
+): DoiResourceTypeGeneral {
+  if (type === EntityType.PROJECT || type === EntityType.FOLDER) {
+    return DoiResourceTypeGeneral.Collection
+  } else {
+    return DoiResourceTypeGeneral.Dataset
+  }
+}
+
+function convertDoiToFormData(doi: V2Doi): DoiFormData {
+  return {
+    titles: (doi.titles || []).map(title => title.title!),
+    creators: (doi.creators || []).map(creator => creator.creatorName!),
+    resourceTypeGeneral: doi.resourceType?.resourceTypeGeneral,
+    publicationYear: doi.publicationYear,
+  }
+}
+
+/**
+ * Returns true if the DOI has fields that are not exposed in the UI, so we can let the user know that any changes
+ * may be destructive to hidden fields.
+ */
+function getWasModifiedViaAPI(doi: V2Doi) {
+  if (doi.creators?.some(creator => !isEmpty(creator.nameIdentifiers))) {
+    return true
+  }
+  return false
+}
+
+function convertFormDataToDoi(formData: DoiFormData): V2Doi {
+  return {
+    titles: formData.titles?.map(title => ({ title })) || [],
+    creators:
+      formData.creators?.map(creator => ({ creatorName: creator })) || [],
+    resourceType: {
+      resourceTypeGeneral: formData.resourceTypeGeneral,
+    },
+    publicationYear: formData.publicationYear,
+  }
+}
+
+export function CreateOrUpdateDoiModal(props: CreateOrUpdateDoiModalProps) {
+  const { open, onClose, objectType, objectId, defaultVersionNumber } = props
+  const [selectedVersionNumber, setSelectedVersionNumber] =
+    useState(defaultVersionNumber)
+
+  const [formData, setFormData] = useState<DoiFormData>({
+    titles: [''],
+    creators: [''],
+  })
+  const { data: currentUser, isLoading: isLoadingCurrentUser } =
+    useGetCurrentUserProfile()
+  const { data: entity, isLoading: isLoadingEntity } = useGetEntity(
+    objectId,
+    defaultVersionNumber,
```

Comment:
any reason to use the selectedVersionNumber for the entity data instead (should they select a different version)?

Reply:
I think that's what the original implementation does -- it would be useful because the title of the entity may have changed between versions, which we could use to populate the form. I will update this.